Explore: Update live tail buttons

aocenas opened the pull request:
Based on: https://www.figma.com/file/fjlhfGs8JYYSYxR0g2occv/Explore-Live-Tailing?node-id=0%3A1

![Kapture 2019-09-16 at 21 00 39](https://user-images.githubusercontent.com/1014802/64985568-71264080-d8c5-11e9-897c-8e26b747c47c.gif)

- Makes run button a segmented with refresh interval.
- Separate Live tail button with segment for pause/resume and stopping.

Things that are not implemented in this PR (want to do that separately):
- Meta labels change
- Change to what is shown after live tail stops

Some design issues that also can be postponed (maybe?):
- The button does not look great imho. Don't like the hover state, active state is missing the blinking state feels a bit odd. In general doing a good button needs a bit more attention and here I am also fighting a bit with existing styles.
- The refresh dropdown looks weird right now, with orange accent and dark background. Also not aligned properly with smaller version of the run button.
- No animation on button state changes.

aocenas committed:
Add new button with functionality

aocenas committed:
Add styling for live button

aocenas committed:
Change time controls in toolbar

aocenas commented on packages/grafana-ui/src/components/RefreshPicker/RefreshPicker.tsx:
```diff
@@ -11,14 +11,22 @@ export const isLive = (refreshInterval: string): boolean => refreshInterval ===
+  // Make it exported as static properties to be easier to access. The global exports need to be accessed by direct
+  // import of this source file which won't work if this was installed as package.
+  static offOption = offOption;
+  static liveOption = liveOption;
```
This should be also refactored elsewhere where the file exports are used.

aocenas commented on public/app/features/explore/ExploreToolbar.tsx:
```diff
@@ -147,6 +130,28 @@ export class UnConnectedExploreToolbar extends PureComponent<Props, {}> {
 
+  stopLive = () => {
+    const { exploreId } = this.props;
+    // TODO referencing this from perspective of refresh picker when there is designated button for it now is not
```
Not sure if it makes more sense here or other PR, but definitely don't won't to leave it like that.

aocenas commented on public/app/features/explore/LiveTailButton.tsx:
```diff
@@ -0,0 +1,104 @@
+import { GrafanaTheme, GrafanaThemeType, useTheme } from '@grafana/ui';
+import tinycolor from 'tinycolor2';
+
+const orangeDark = '#FF780A';
```
@dprokop Fumbling with the colors here. Not sure if I should just put them in the theme or not. `FF780A` is used but only in color picker.

dprokop replied:
@aocenas I would suggest using a color from the palette that's the closes one instead of introducing a new one. @Ijin08 any suggestions/?

aocenas replied:
@dprokop how would I do that? I mean should I just export it and import here and use the one I need? There is also the other shade which is new so not sure whether it make sense to reuse just for this one, seems like keeping it as an exception is better for now.

torkelo commented:
<img width="284" alt="Screen Shot 2019-09-17 at 07 09 16" src="https://user-images.githubusercontent.com/10999/65013143-22ef5c80-d91a-11e9-9d68-2ae8a272e6e8.png">

The orange text on blue is not great :) 

The enter key icon to me feels scrupulous but not sure what @davkal thoughts are on the enter key icon. 

Otherwise I think the new live / pause / stop buttons work really well and good. 

But I find the buttons at the end of the live tail logs view should looks similar? 
<img width="460" alt="Screen Shot 2019-09-17 at 07 11 58" src="https://user-images.githubusercontent.com/10999/65013213-87122080-d91a-11e9-9ab0-799fb3d477fa.png">

dprokop commented:
I find it quite annoying when starting the live tailing makes the rest of the toolbar buttons to disappear, resulting in jumpy interface. Why not disabling the rest of the buttons instead?

jessover9000 commented:
> The enter key icon to me feels scrupulous but not sure what @davkal thoughts are on the enter key icon.

@aocenas if you check the mockups, the Enter icon is supposed to be replaced by refresh ☝️ (I think it was this one https://fontawesome.com/icons/sync?style=solid)

General adjustments:
1. since the Live tailing switches the view/mode, I think it's better to give it some margin-left to set it apart from the rest of the controls. I recommend 12px.
2. the text "Run query" could use some margin-left so that the icon/spinner isn't just as close to it
3. the pause flashing of the "live tailing" can be half as often. also, I feel like white is too vibrant a color here – maybe flash between orange and #8e8e8e (as the rest of secondary button text) and only go lighter on hover?
4. in live mode when the split button of "Run query" disappears, the button corners on the right are less round than on the left ;) #nitpicking
5. I feel like the transition between "in live mode" and "normal mode" is quite rough at this point. not sure what's possible here but can we do a crossfading effect for the entire viewport or something? just to soften the change a bit.

@dprokop I think having several disabled buttons in the toolbar there is a waste of space and not all that helpful – since we're losing the dropdown of the Run query buttons, the buttons will jump either way. I think a better transition might solve that issue :)

jessover9000 commented:
> But I find the buttons at the end of the live tail logs view should looks similar?

Well we should definitely not make the exact same orange button appear twice on the screen with a flashing pause state etc., I think that would be too much visually. I guess we could do a split button. Not sure whether primary or secondary though 🤔 Feel like another blue split button down there might be too much, but I like how the Play/Pause is currently more vibrant than just grey.

dprokop commented:
> @dprokop I think having several disabled buttons in the toolbar there is a waste of space and not all that helpful – since we're losing the dropdown of the Run query buttons, the buttons will jump either way. I think a better transition might solve that issue :)

Don't necessarily agree as those buttons are there in the first place as a consequence of an initial state. And since the `Run query` button is non-functional when live tailing(isn't it?) it could also be disabled, with caret visible. Not also sure if it should be a button at all if it only shows loading indicator. The only reason for jumping would be the ⬛button in added to live tailing button.

aocenas committed:
Change select button color

aocenas committed:
Change icon

aocenas commented:
- Fixed the refresh value color from orange to white.
- Changed the icon to refresh
<img width="477" alt="Screenshot 2019-09-17 at 10 50 52" src="https://user-images.githubusercontent.com/1014802/65026543-0e6e8c80-d939-11e9-9704-8702339b3cc2.png">
Not sure if this is that clear with collapsed button right now. @davkal @jessover9000

aocenas commented on public/app/features/explore/ExploreToolbar.tsx:
```diff
@@ -147,6 +130,28 @@ export class UnConnectedExploreToolbar extends PureComponent<Props, {}> {
+    const { exploreId } = this.props;
+    // TODO referencing this from perspective of refresh picker when there is designated button for it now is not
+    //  great. Needs another refactor.
+    this.props.changeRefreshIntervalAction({ exploreId, refreshInterval: (RefreshPicker as any).offOption.value });
```
@dprokop Got myself into a pickle here. I wanted to reference these options as statics on `RefreshPicker` for easier import. That works fine until I wrapped it into `withTheme`. HoistNonReactStatics works but the types are lost and typescript does not know that it still has those statics. Do you know if there is a solution for that?

dprokop replied:
sorry, don't have ready solution for this:(

aocenas replied:
Fixed it some extent in withTheme adding another generic param to specify the resulting statics. Not great means you need to manually define your statics again:
```typescript
export const RefreshPicker = withTheme<
  Props,
  {
    offOption: typeof RefreshPickerBase.offOption;
    liveOption: typeof RefreshPickerBase.liveOption;
  }
>(RefreshPickerBase);
```
But I assume it is fine for the rare occasions if we do not have better solution.

aocenas committed:
Fix statics in WithTheme

davkal reviewed (approved):
LGTM

aocenas commented:
Merged, will create an issue and separately work on the other UI improvements.
